File: leetcode/0102-binary-tree-level-order-traversal/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        traversal = list()
        
        if root == None:
            return traversal
        
        q = deque()
        q.append(root)
        q.append(None)
        
        level = 0
        traversal.append(list())
        
        while len(q) != 0:
            node = q.popleft()
            
            if (node == None):
                level += 1
                
                if len(q) != 0:
                    q.append(None)
                    traversal.append(list())
                
                continue
            
            if node.left:
                q.append(node.left)
            if node.right:
                q.append(node.right)
                
            traversal[level].append(node.val)
        
        return traversal

File: leetcode/0102-binary-tree-level-order-traversal/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> traversal;
        
        if (root == NULL)
            return traversal;
        
        queue<TreeNode*> q;
        q.push(root);
        q.push(NULL);
        int level = 0;
        vector<int> tmp1;
        traversal.push_back(tmp1);
        
        while (!q.empty()) {
            TreeNode* node = q.front();
            q.pop();
            
            if (node == NULL) {
                level++;
                
                if (!q.empty()) {
                    q.push(NULL);
                    vector<int> tmp2;
                    traversal.push_back(tmp2);
                }
                
                continue;
            }
            
            if (node->left != NULL)
                q.push(node->left);
            if (node->right != NULL)
                q.push(node->right);
            
            traversal[level].push_back(node->val);
        }
        
        return traversal;
    }
};